### ISSUE TITLE
ReactorBuilder.runReactor should cancel the timeout it creates if it wasn't triggered

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5634]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5634|
|Type|enhancement|
|Created|2012-04-11 01:43:31Z|
|Branch|https://github.com/twisted/twisted/tree/reactorbuilder-runreactor-timeout-5634|

`twisted.internet.test.test_tcp.AbortConnectionMixin` needs to manually deal with the timeout added by `ReactorBuilder.runReactor`. It would be better if `runReactor` cleaned up after itself and cancelled the timeout it scheduled if it exists before the timeout is reached.

Attachments:

* [5634.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7d6/7d63ec400bb65da3554e846a2d40a9a1ffb336be/0b9790e667409a639ec63709ac8df91c8897ca3c.patch) (1299 bytes) - added by Saurabh on 2013-10-14 02:15:14Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5634 5634
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_reactorbuilder_runreactor_timeout_5634 branches/reactorbuilder-runreactor-timeout-5634
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1334108611000000 1334108611000000
changetime__1381839820000000 1381839820000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Saurabh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Saurabh set owner to Saurabh|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Saurabh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Saurabh removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [40331]) Branching to 'reactorbuilder-runreactor-timeout-5634'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [40332]) apply 5634.patch 

refs [#5634](https://github.com/twisted/twisted/issues/5634)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.

It looks like the new `if` could just be an `else` on the `if` that's there already.  Also, the change needs a [news fragment](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) - but just a misc, I expect.

I'll make these two minor changes and apply.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Oh.  Fortunately before I merged I also noticed that the change to the tests is wrong.  The assertion near the comment should be changed - not deleted.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [40335]) Restore the assertion about delayed calls in the reactor (or lack thereof).

refs [#5634](https://github.com/twisted/twisted/issues/5634)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [40340]) Merge reactorbuilder-runreactor-timeout-5634 

Author: Saurabh
Reviewer: exarkun
Fixes: [#5634](https://github.com/twisted/twisted/issues/5634)

Cancel the timeout set up by `ReactorBuilder.runReactor` and remove
the work-around in a TCP unit test that previously accounted for
this extra delayed call left lying around.
